File: TypeScript/binary-search.ts
/* -------------------------------------------------------------------------------|
 |                                                                                 |
 |                     Binary Search Algorithm                                     |
 |                                                                                 |
 | --------------------------------------------------------------------------------|
 | Created By   ### Ugur Ayan  ###                                                 |
 | Created Date ### 13/10/2017 ###                                                 |
 | www.ugurayan.com.tr                                                             |
 | ugur.ayan@ugurayan.com.tr                                                       |
 | ------------------------------------------------------------------------------- |
 | Definition:                                                                     |
 | Given a sorted array arr[] of n elements, write a function to search a given    |
 | element x in arr[].                                                             |
 | ------------------------------------------------------------------------------- |
 | Complexity : O(Log n)                                                           |
 | -------------------------------------------------------------------------------*/

var A, num;

binarySearch(A, num) {
    var first   = 1;
    var last    = n;
    var flag    = 0;
    var mid     = 0;
    var index   = -1;
    const n     = A.lengthl;

    while (first <= last) {
        mid = Math.ceil((first + last) / 2);

        if (A[mid] === num) {
            index = mid;
            flag = 1;
            break;
        } else if (A[mid] > num) {
            last = mid - 1;
        } else
            first = mid + 1;
    }

    if (flag == 0) index = -1;
}

File: Matlab/binary-search.m
function [index] = binary-search(A, n, num)
%-------------------------------------------------------------------------
%
%                     Binary Search Algorithm
%
% ------------------------------------------------------------------------
% Created By   ### Ugur Ayan  ###
% Created Date ### 16/10/2017 ###
% www.ugurayan.com.tr
% ugur.ayan@ugurayan.com.tr
% ------------------------------------------------------------------------
% Syntax:       [index] = binarySearch(A, n, num);
%
% Inputs:       A: Array (sorted) that you want to search
%               n: Length of array A
%               num: Number you want to search in array A
%
% Outputs:      index: Return position in A that A(index) == num
%                      or -1 if num does not exist in A
%
% Definition:  This function find number in array (sorted) using binary
%               search
%
% ------------------------------------------------------------------------
% Complexity : O(1)          best-case performance
%            : O(log2(n))    worst-case performance%
%            : O(1)          auxiliary space
% ------------------------------------------------------------------------


%
%

    first = 1;
    last = n;
    flag = 0;

    while first <= last
        mid = ceil((first + last) / 2);

        if A(mid) == num
            index = mid;
            flag = 1;
            break;
        else if A(mid) > num
            last = mid - 1;
            else
                first = mid + 1;
            end
        end
    end

    if flag == 0;
        index = -1;
    end

end
